nengo_bones/config.py

```diff
@@ -0,0 +1,114 @@
+    -------
+    filename : str
+        File path for default config file.
+    """
+    # for now, assume that config file is in cwd
+    conf_file = os.path.join(os.getcwd(), ".nengobones.yml")
+
+    return conf_file
+
+
+def fill_defaults(config):
+    """
+    Fills in default values in a loaded config (in-place).
+
+    Parameters
+    ----------
+    config : dict
+        Dictionary containing configuration values.
+    """
+    if "travis_config" in config:
+        config["travis_config"].setdefault("python_version", "3.6")
+        config["travis_config"].setdefault("global_vars", {})
+        config["travis_config"].setdefault("pypi_user", None)
+        config["travis_config"].setdefault("deploy_dists", ["sdist"])
+
+    if "codecov_config" in config:
+        config["codecov_config"].setdefault("skip_appveyor", True)
+        config["codecov_config"].setdefault("abs_target", "auto")
+        config["codecov_config"].setdefault("diff_target", "100%")
+
+
+def validate_config(config):
+    """
+    Apply validation to a populated config dict.
+
+    Parameters
+    ----------
+    config : dict
+        Dictionary containing configuration values.
+    """
+    mandatory = ["pkg_name", "repo_name"]
+    if "travis_config" in config:
+        mandatory.append("travis_config.jobs")
+
+    for entry in mandatory:
+        tmp = config
+        for key in entry.split("."):
+            try:
+                tmp = tmp[key]
+            except KeyError:
+                raise KeyError("Config file must define %s" % entry)
+
+    if "script_config" in config:
+        for cfg in config["script_config"]:
+            if "template" not in cfg:
+                raise KeyError("Script config must define 'template' "
+                               "(for entry %s)" % cfg)
+
+            try:
+                # make sure that people don't accidentally do
+                # pip_install: dependency (which gives a string), rather than
+                # pip_install:
+                #   - dependency
+                for key in ("pip_install", "conda_install"):
+                    if not isinstance(cfg[key], list):
+                        raise TypeError(
+                            "%s should be a list, found '%s'; did you forget "
+                            "to add '-' before each dependency?" % (
+                                key, cfg[key]))
+            except KeyError:
+                pass
+
+    # TODO: check that there aren't unused config options in yml
+
+
+def load_config(conf_file=None):
+    """
+    Load config values from file and apply defaults/validation.
+
+    Parameters
+    ----------
+    conf_file : str
+        Filepath for config file (if None, will load the default returned by
+        `.find_config`.
+
+    Returns
+    -------
+    config : dict
+        Dictionary containing configuration values.
+    """
+
+    if conf_file is None:
+        conf_file = find_config()
+
+    with open(str(conf_file)) as f:
```

Comment:
Just curious why you wrap this in `str()`?

Reply:
In case `conf_file` is a path object (e.g. https://docs.python.org/3/library/pathlib.html or https://py.readthedocs.io/en/latest/path.html), rather than an actual string.  For example in the unit tests we use the `tmpdir` pytest fixture, which returns a py.path object.  In Python 3.6 you can use those directly with `open`, but in `<=3.5` you need to explicitly convert them to strings.